include/picongpu/particles/shapes/P4S.hpp

```diff
@@ -21,135 +21,146 @@
         {
-            /*       -
-             *       |  115/192 + x^2 * (-5/8 + 1/4 * x^2)                          if -1/2 < x < 1/2
-             * W(x)=<|
-             *       |  1/96 * (55 + 4 * x * (5 - 2 * x * (15 + 2 * x * (-5 + x)))) if 1/2 <= |x| < 3/2
-             *       |
-             *       |  1/384 * (5 - 2 * x)^4                                       if 3/2 <= |x| < 5/2
-             *       -
+            /*
+             * W(x)= 115/192 - 5/8 * x^2 + 1/4 * x^4
+             *     = 115/192 + x^2 * (-5/8 + 1/4 * x^2)
              */
-            float_X const abs_x = math::abs( x );
-
-            bool const below_2nd_radius = abs_x < 1.5_X;
-            bool const below_1st_radius = abs_x < 0.5_X;
+            float_X const square_x = x * x;
+            return 115._X / 192._X + square_x * (
+                -5._X / 8._X +
+                1.0_X / 4.0_X * square_x
+            );
+        }
 
-            float_X const rad1 = ff_1st_radius( abs_x );
-            float_X const rad2 = ff_2nd_radius( abs_x );
-            float_X const rad3 = ff_3rd_radius( abs_x );
+        HDINLINE static float_X ff_2nd_radius( float_X const x )
+        {
+            /*
+             * W(x)= 1/96 * (55 + 20 * x - 120 * x^2 + 80 * x^3 - 16 * x^4)
+             *     = 1/96 * (55 + 4 * x * (5 - 2 * x * (15 + 2 * x * (-5 + x))))
+             */
+            return 1._X / 96._X * (
+                55._X + 4._X * x * (
+                    5._X - 2._X * x * (
+                        15._X + 2._X * x * (
+                            -5._X + x
+                        )
+                    )
+                )
+            );
+        }
 
-            float_X result = rad3;
-            if( below_1st_radius )
-                result = rad1;
-            else if( below_2nd_radius )
-                result = rad2;
+        HDINLINE static float_X ff_3rd_radius( float_X const x )
+        {
+            /*
+             * W(x)=1/384 * (5 - 2*x)^4
+             */
+            float_X const tmp = 5._X - 2._X * x;
+            float_X const square_tmp = tmp * tmp;
+            float_X const biquadratic_tmp = square_tmp * square_tmp;
 
-            return result;
+            return 1._X / 384._X * biquadratic_tmp;
         }
-
     };
 
-    struct ChargeAssignment : public shared_P4S::P4S
+} // namespace detail
+
+    /** Piecewise cubic cloud particle shape
+     *
+     * Note that this shape name does not follow the NGP / CIC / TSC notation
+     * which names the cloud density. Rather, P4S is the name of the
+     * corresponding assignment function, same as for the PCS shape.
+     *
+     * Cloud density form: piecewise cubic
+     * Assignment function: fourth order B-spline
+     */
+    struct P4S
     {
 
-        HDINLINE float_X operator()( float_X const x )
+        //! Order of the assignment function spline
+        static constexpr uint32_t assignmentFunctionOrder = 4u;
```

Comment:
I think it would be better to define this as 
```suggestion
        static constexpr uint32_t assignmentFunctionOrder = support - 1;
```

Reply:
It is how that was effectively used before. This breaks for the `Counter` shape which would result in -1 order, as it has `support` of 0. Now, I am not sure why is it implemented like that, since its actual support is 1, but it seems now the implementation of particle counting depends on that.

Reply:
Perhaps it is reasonable to make it with this type of equation in every shape but `Counter`.